daemon/settings.c

```diff
@@ -1159,8 +1173,43 @@ appsettings_encode(const appsettings_t *self, GKeyFile *file)
 static void
 appsettings_rethink(appsettings_t *self)
 {
-    /* Assign current value -> masking is re-applied */
-    appsettings_set_granted(self, self->ast_granted);
+    log_info("%s(uid=%d): rethink",
+             appsettings_appname(self),
+             appsettings_uid(self));
+
+    int permission_change = appsettings_update_permissions(self);
+
+    const stringset_t *permissions = appsettings_get_permissions(self);
+    const stringset_t *granted = appsettings_get_granted(self);
+
+    if( appsettings_update_autogrant(self, appsettings_get_allowlisted(self)) ) {
+        /* Autogrant config changed: choose all or nothing */
+        if( appsettings_get_allowed(self) != APP_ALLOWED_NEVER ) {
+            appsettings_update_allowed(self, APP_ALLOWED_UNSET);
+            granted = permissions;
+        }
+    }
+    else {
+        /* Apply autogrant policy */
+        switch( appsettings_get_autogrant(self) ) {
+        case APP_GRANT_ALWAYS:
+            /* Keep in sync with application requirements */
+            granted = permissions;
+            break;
+        case APP_GRANT_LAUNCH:
+            /* Sync if new permissions are required */
+            if( permission_change > 0 )
+                granted = permissions;
+            break;
+        default:
+            /* Prompt user if new permissions are required */
+            if( permission_change > 0 )
+                appsettings_update_allowed(self, APP_ALLOWED_UNSET);
```

Comment:
If allowed was NEVER, should this be changing it?

Reply:
A bit grey area... but now that you brought it up I would perhaps keep the denied state if user has gone through the trouble to get there - will fix.

Reply:
done